---
title: "Bash commands run on NeSI"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Filter initial assembly by best supported contig list and output representative contig FASTA file
```{console}
$ module load seqtk/1.3-gimkl-2018b

$ seqtk subseq Trinity_taxon_x.fasta taxon_x_bestsupported_contigs_list.txt > ./taxon_x_trinity_rep_contigs.fasta
```

Append assembly identifier to TransDecoder peptide (.pep) and nucleotide (.cds files)
```{console}
$ module load bioawk/1.0

$ # .pep
$ time bioawk -c fastx '{ print ">assembly_ID_" $name "\n" $seq }' high_expression_contigs_taxon.fasta.transdecoder.pep > assembly_ID_id.faa
$ # .cds
$ time bioawk -c fastx '{ print ">assembly_ID_" $name "\n" $seq }' high_expression_contigs_taxon.fasta.transdecoder.cds > assembly_ID_id.fna
```

Count unique R. crithmifolius and R. monroi-Mt Hutt BLASTp hits against Arabidopsis protein sequences 
```{console}
$ # Isoforms
$ cut -f 2 blastp_taxon_x_arabidopsis.outfmt6 | sort | uniq | wc -l

$ # Genes (remove last 2 characters from the AT isoform ID)
$ cut -f 2 blastp_taxon_x_arabidopsis.outfmt6 | sed -r 's/.{2}$//' | sort | uniq | wc -l
```



```{console}
$ 
$ 

$ 
```



```{console}
$ 
$ 

$ 
```



```{console}
$ 
$ 

$ 
```



```{console}
$ 
$ 

$ 
```



```{console}
$ 
$ 

$ 
```



```{console}
$ 
$ 

$ 
```
